File: 904. Fruit Into Baskets/slidewindow.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> cap;
        int res = 0;
        int left = 0;
        cap[fruits[0]]++;

        if (fruits.size() == 1)
            return 1;

        for (int i = 1; i < fruits.size(); i++) {
            cap[fruits[i]]++;
            while (cap.size() > 2) {
                cap[fruits[left]]--;
                if (cap[fruits[left]] == 0) {
                    cap.erase(fruits[left]);
                }
                left++;
            }
            res = max(res, i - left + 1);
        }
        return res;
    }
};

File: 904. Fruit Into Baskets/slidewindow.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        cap = {}
        res = 0
        left = 0
        cap[fruits[0]] = 1

        if len(fruits) == 1:
            return 1

        for i in range(1, len(fruits)):
            cap[fruits[i]] = cap.get(fruits[i], 0) + 1
            while len(cap) > 2:
                cap[fruits[left]] -= 1
                if cap[fruits[left]] == 0:
                    cap.pop(fruits[left])
                left += 1
            res = max(res, i - left + 1)

        return res
